frontend/src/metabase/common/components/EntityPicker/components/NestedItemPicker/NestedItemPicker.tsx

```diff
@@ -0,0 +1,87 @@
+import type React from "react";
+
+import ErrorBoundary from "metabase/ErrorBoundary";
+import { Flex } from "metabase/ui";
+import type { SearchListQuery } from "metabase-types/api";
+
+import type {
+  EntityPickerOptions,
+  TypeWithModel,
+  PickerState,
+  TisFolder,
+} from "../../types";
+import type { EntityItemListProps } from "../ItemList";
+
+import { AutoScrollBox } from "./AutoScrollBox";
+import { ListBox } from "./NestedItemPicker.styled";
+
+export interface NestedItemPickerProps<TItem extends TypeWithModel> {
+  onFolderSelect: ({ folder }: { folder: TItem }) => void;
+  onItemSelect: (item: TItem) => void;
+  itemName: string;
+  options: EntityPickerOptions;
+  path: PickerState<TItem>;
+  isFolder: TisFolder<TItem>;
+  listResolver: React.FC<
+    EntityItemListProps<TItem> & {
+      options: EntityPickerOptions;
+    }
+  >;
+}
+
+const generateKey = (query?: SearchListQuery) =>
+  JSON.stringify(query ?? "root").slice(0, 255);
```

Comment:
!nit we should use a proper hashing function

Reply:
I wondered about collisions here. If two queries that are meant to be distinct end up using the same key, could that cause problems?

Reply:
Let me think about this... there are native hashing function, but I'm not 100% confident about browser support:
https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest

Reply:
looks like that would fail on http:// origins https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto

Reply:
> I think collision is basically impossibly with our current API
If that's the case with 255 characters, let's just make it literally impossible and not slice the JSON output?

Reply:
Sorry, I should have been more judicious about _why_ this was needed. Can't we rely on object equality? It looks like this has is ultimately just being used by the auto-scrolling feature to detect a change. Couldn't we pass down `path[path.length - 1].query` instead of `generateKey(path[path.length - 1].query)`?